server/cluster/unsafe_recovery_controller.go

```diff
@@ -125,12 +128,216 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
 }
 
 func (u *unsafeRecoveryController) isPlanExecuted(report *pdpb.StoreReport) bool {
-	return true
+	targetRegions := make(map[uint64]*metapb.Region)
+	toBeRemovedRegions := make(map[uint64]bool)
+	storeId := report.StoreId
+	for _, create := range u.storeRecoveryPlans[storeId].Creates {
+	        targetRegions[create.Id] = create
+	}
+	for _, update := range u.storeRecoveryPlans[storeId].Updates {
+	        targetRegions[update.Id] = update
+	}
+	for _, del := range u.storeRecoveryPlans[storeId].Deletes {
+	        toBeRemovedRegions[del] = true
+	}
+	numFinished := 0
+	for _, peerReport := range report.Reports {
+		region := peerReport.RegionState.Region
+		if _, ok := toBeRemovedRegions[region.Id]; ok {
+			return false
+		} else if target, ok := targetRegions[region.Id]; ok {
+			if bytes.Equal(target.StartKey, region.StartKey) && bytes.Equal(target.EndKey, region.EndKey) {
+				numFinished += 1
+			} else {
+				return false
+			}
+		}
+	}
+	return numFinished == len(targetRegions)
+
+}
+
+type regionItem struct {
+	region *metapb.Region
+}
+
+func (r regionItem) Less(other btree.Item) bool {
+	return bytes.Compare(r.region.StartKey, other.(regionItem).region.StartKey) < 0
+}
+
+func (u *unsafeRecoveryController) canElectLeader(region *metapb.Region) bool {
+	numFailedPeers := 0
+	for _, peer := range region.Peers {
+		if _, ok := u.failedStores[peer.StoreId]; ok {
+			numFailedPeers += 1
+		}
+	}
+	return numFailedPeers*2 < len(region.Peers)
+}
+
+func (u *unsafeRecoveryController) removeFailedStores(region *metapb.Region) bool {
+	updated := false
+	var newPeerList []*metapb.Peer
+	for _, peer := range region.Peers {
+		if _, ok := u.failedStores[peer.StoreId]; !ok {
+			newPeerList = append(newPeerList, peer)
+		} else {
+			updated = true
+		}
+	}
+	region.Peers = newPeerList
+	return updated
+}
+
+type peerStorePair struct {
+	peer    *pdpb.PeerReport
+	storeId uint64
 }
 
 func (u *unsafeRecoveryController) generateRecoveryPlan() {
 	u.Lock()
 	defer u.Unlock()
+	healthyRegions := make(map[uint64]*pdpb.PeerReport)
+	var allPeerReports []*peerStorePair
+	for storeId, storeReport := range u.storeReports {
+		for _, peerReport := range storeReport.Reports {
+			allPeerReports = append(allPeerReports, &peerStorePair{peerReport, storeId})
+			regionId := peerReport.RegionState.Region.Id
+			if existing, ok := healthyRegions[regionId]; ok && existing.RaftState.LastIndex > peerReport.RaftState.LastIndex {
+				// Only keeps the newest region report.
+				continue
+			}
+			if u.canElectLeader(peerReport.RegionState.Region) {
+				healthyRegions[regionId] = peerReport
+			} else {
+				delete(healthyRegions, regionId)
```

Comment:
A is a healthy region, right? I think the current algorithm gives the correct solution, which is keep both A and B, only removes 4,5 from the peer list?

Reply:
A is not a healthy region.

Reply:
Let's assume the peer reports come at the sequence of A=[a,c) from store 2, A=[a,b) from and A=[a,c) from store 3. Then the final peer report left in `healthyRegions` will be A=[a,c) from store 3 which is a stale peer info.

Reply:
As we discussed offline, the safest solution is to let all the TiKV instances stop conf changes and merge/split operations and wait for the pending ones to finish before sending the report to PD, otherwise, there is no guarantee that PD receives the up-to-date region ranges. This is going to be support in phase 2, and currently we assume that, among all the peer reports from the same region, at least one of them is up-to-date (although it is not always the case).